Update library to latest webgpu-native headers

PJB3005 opened the pull request:
This goes up to https://github.com/webgpu-native/webgpu-headers/commit/6f549cc3409663303b06f50483b9ebf9d1cf19dd
Things I *didn't* do:

* I didn't update the library to make sure "instance dropped" callback error codes are guaranteed to happen, like they seem to be in Dawn.

List of changes (roughly in order of header commits):

* Various enum and struct renames
* Updated callbacks to use the new `*CallbackInfo` structs and 2-userdata system. Also updated functions to return `WGPUFuture`, though the `WGPUFuture` thing is just stubbed out at the moment as I don't think wgpu-core has the necessary functionality for it. `wgpuInstanceWaitAny` is `unimplemented!()`
* `DepthClipControl` merged into `PrimitiveState`, related code simplified.
* Updated `depthWriteEnabled` to use an optional bool, mostly matters due to added validation.
* Add TODOs for missing features (sliced 3D compressed textures)
* `*Reference()` became `*AddRef()`
* Added `unorm10-10-10-2` vertex format
* Usage field in `TextureViewDescriptor`, just used for validation as wgpu-core doesn't allow specifying it anyways.
* Removed `maxInterStageShaderComponents`
* Added `clang_macro_fallback()` to bindgen config, since the headers switched to using `UINT32_MAX` etc. `UINT64_MAX` still doesn't work so I had to manually define those.
* Renamed flags enums. Added a conversion helper function to convert them from u64 -> u32 for mapping. (means added direct dependency on bitflags crate)
* Removed device argument from (unimplemented) `wgpuGetProcAddress`
* Suboptimal surface texture acquisition moved to enum return value, was easy since wgpu-core already returns it like that.
* "Undefined" present mode added, it just selects FIFO.
* `*EnumerateFeatures()` has been replaced with `*GetFeatures()`.
* All strings are now passed as `WGPUStringView`.

PJB3005 committed:
Update library to latest webgpu-native headers

This goes up to https://github.com/webgpu-native/webgpu-headers/commit/2b5974750be7b2545d842ecec02e585022081952

Things I *didn't* do:

* I didn't update the library to make sure "instance dropped" callback
  error codes are guaranteed to happen, like they seem to be in Dawn.

List of changes (roughly in order of header commits):

Various enum and struct renames

Updated callbacks to use the new *CallbackInfo structs and 2-userdata system. Also updated functions to return WGPUFuture, though the WGPUFuture thing is just stubbed out at the moment as I don't think wgpu-core has the necessary functionality for it. wgpuInstanceWaitAny is unimplemented!()

DepthClipControl merged into PrimitiveState, related code simplified.

Updated depthWriteEnabled to use an optional bool, mostly matters due to added validation.

Add TODOs for missing features (sliced 3D compressed textures)

*Reference() became *AddRef()

Added unorm10-10-10-2 vertex format

Usage field in TextureViewDescriptor, just used for validation as wgpu-core doesn't allow specifying it anyways.

Removed maxInterStageShaderComponents

Added clang_macro_fallback to bindgen config, since the headers switched to using UINT32_MAX etc. UINT64_MAX still doesn't work so I had to manually define those.

Renamed flags enums. Added a conversion helper function to convert them from u64 -> u32 for mapping. (means added direct dependency on bitflags crate)

Removed device argument from (unimplemented) wgpuGetProcAddress

Suboptimal surface texture acquisition moved to enum return value, was easy since wgpu-core already returns it like that.

"Undefined" present mode added, it just selects FIFO.

PJB3005 committed:
Update C examples to new headers

PJB3005 committed:
Fix WGPUBufferMapAsyncStatus tomfoolery

This enum was replaced with WGPUMapAsyncStatus, but I didn't quite notice. The error codes map different.

PJB3005 committed:
Fix macOS typos in examples

PJB3005 commented:
Remaining Windows CI failure is caused by https://github.com/webgpu-native/webgpu-headers/issues/339

PJB3005 committed:
Update headers again

Replaces *EnumerateFeatures with *GetFeatures.

Also fixes CI due to fix in headers.

almarklein reviewed (commented):
LGTM but maybe @rajveermalviya can have a look at the Rust code.

PJB3005 committed:
Update wgpu.h to match new flag scheme

Upstream removed the "Flags" suffix from flags types and moved them to no longer be C enums. This matches that change. WGPUInstanceFlag still has "Flag" in the name because, well, there'd be nothing left to distinguish it from WGPUInstance, and it makes sense for it.

eliemichel commented on src/conv.rs:
```diff
@@ -1660,13 +1648,18 @@ pub fn map_surface_configuration(
+    let present_mode = match config.presentMode {
+        native::WGPUPresentMode_Undefined => wgt::PresentMode::Fifo,
+        _ => map_present_mode(config.presentMode),
+    };
```
I believe this `match` should be part of `map_present_mode` rather than being added on top of it

PJB3005 replied:
I did this here because I wasn't sure if I could cleanly add that to `map_present_mode`, as it's defined with a `map_enum!` macro.

PJB3005 replied:
Ended up doing this properly due to more incoming changes in the latest headers.

eliemichel commented on src/conv.rs:
```diff
@@ -1701,3 +1694,11 @@ pub fn map_adapter_type(device_type: wgt::DeviceType) -> native::WGPUAdapterType
     }
 }
+
+pub fn from_u64_bits<T: bitflags::Flags<Bits = u32>>(value: u64) -> Option<T> {
```
(Note: Ideally bitfields would switch to `u64` in upstream `wgpu` instead of needing a conversion here)

eliemichel commented on Cargo.toml:
```diff
@@ -157,6 +157,7 @@ log = "0.4"
 thiserror = "1"
 parking_lot = "0.12"
 smallvec = "1"
+bitflags = "2.6.0"
```
why is this needed?

PJB3005 replied:
It's to be able to refer to `bitflags` types for the `from_u64_bits` helper.

eliemichel reviewed (approved):
Nice job! I started looking at how to implement futures in `wgpu` a couple of days ago, nothing obvious but I'll share if I figure sth out!

PJB3005 committed:
Update headers again, WGPUStringView

Also updates wgpu.h to use WGPUStringView everywhere.

PJB3005 committed:
Only specify major version for bitflags dependency

PJB3005 committed:
Remove redundant unsafe block

eliemichel commented:
For WGPUFuture, we need to have access to submission index (and return that in the future's id). Here is the beginning of an attempt at adding this upstream: https://github.com/gfx-rs/wgpu/pull/6360

PJB3005 commented on src/utils.rs:
```diff
@@ -268,6 +270,69 @@ macro_rules! map_enum {
+            _ => make_slice(string_view.data as *const u8, string_view.length),
+        };
+
+        Some(std::str::from_utf8_unchecked(bytes))
```
Important to note: this doesn't check whether strings passed into the library are valid UTF-8. I previously many locations did, but I figured this wasn't necessary.

eliemichel commented:
FYI I started trying to build upon this PR and upgrade wgpu to my [PR that adds submission indices for futures](https://github.com/gfx-rs/wgpu/pull/6360), WIP is here: https://github.com/eliemichel/wgpu-native/tree/eliemichel/2024-10-10-upgrade-wgpu-futures

Ongoing work: I need to figure out what replaced the `DynComputePass` and `DynRenderPass` types in newest wgpu.

ygdrasil-io commented:
There is a missing fonction `wgpuGetInstanceFeatures` on the rust code but describe on yaml as :
```yaml
  - name: get_instance_features
    doc: Query the supported instance features
    args:
      - name: features
        doc: The supported instance features
        type: struct.instance_features
        pointer: mutable
```

eliemichel commented:
> FYI I started trying to build upon this PR and upgrade wgpu

This is now ready, PR: #441

PJB3005 committed:
Add wgpuGetInstanceFeatures stuff

All stubs, since we don't have WaitAny at the moment.

PJB3005 commented:
@ygdrasil-io done, thanks for pointing that out. Though note it's not very useful since WaitAny isn't implemented yet.

PJB3005 committed:
I forgot to cargo fmt again

PJB3005 committed:
Implement "NotUsed" bind group entry types

Updates to https://github.com/webgpu-native/webgpu-headers/commit/6a23100d61ee768f42a81880a0f5c982cffc4c30

PJB3005 committed:
Update to new undefined enums

Update to https://github.com/webgpu-native/webgpu-headers/commit/f1cdc3fd149cebe166991310b34259f9af5b9df5

Also went through a bunch of existing enum conversions and fixed up some seemingly spec-incorrect cases of undefined enums not being handled properly.

As part of this, I made a new helper map_enum_with_undefined!() which distinguishes undefined and unknown enum values. Previously, much code relied on undefined being caught in the same net as an unknown value. This is no longer the case.

PJB3005 committed:
Update for other enum changes

This updates the headers to https://github.com/webgpu-native/webgpu-headers/commit/af63d3476bb9a97c3ac581c21a3273d47027e40b

These changes are exclusively in the header enum values, so no Rust code needs changing. Making this a separate commit for easier review.

PJB3005 committed:
Update for new header features

Update headers to https://github.com/webgpu-native/webgpu-headers/commit/b7656d0ce59a6379b31a58abae956873364e9a17

Adds dual source blending. Other two features (float32 blendable and clip distances feature in wgsl) are not supported by wgpu.

Also made map_blend_factor use the shorter macro form, as all the enum names match (they didn't 3 years ago when this code was originally written, according to Git history).

PJB3005 committed:
Fix getFeatures chained struct handling.

Oops that's not how these chained structs work.

PJB3005 committed:
Flatten limits structures

Updates headers to https://github.com/webgpu-native/webgpu-headers/commit/6f549cc3409663303b06f50483b9ebf9d1cf19dd

Also made matching changes to wgpu.h

PJB3005 committed:
Merge remote-tracking branch 'upstream/trunk' into 24-09-19-update-headers

PJB3005 committed:
Some day I will learn to run cargo fmt and check compile before merges

eliemichel commented:
@rajveermalviya Any specific reason why this PR is on hold? (You not having had time to review it yet is a perfectly fine reason, just trying to figure out if there is a technical blocker I could help with!)

eliemichel commented:
Merged latest commits into #441 and did some extra upgrade of webgpu.h, leading to #445, still hoping for this to get merged!

PJB3005 commented:
Closing since the other PR was merged.
